code/game/objects/items/weapons/RCD.dm

```diff
@@ -382,3 +389,285 @@ RCD
 	name = "admin RCD"
 	max_matter = INFINITY
 	matter = INFINITY
+
+
+/obj/item/weapon/rcd/arcd
+	name = "advanced rapid-construction-device (ARCD)"
+	max_matter = 300
+	matter = 300
+	color = "red"
+
+
+// Ranged RCD
+
+
+/obj/item/weapon/rcd/arcd/afterattack(atom/A, mob/user, proximity)
+	proximity = 1
+	if(!(A in view(7, get_turf(user))))
+		to_chat(user, "<span class='warning'>The \'Out of Range\' light on the RCD blinks red.</span>")
+		return 0
+	if(isturf(A) || istype(A, /obj/machinery/door/airlock) || istype(A, /obj/structure/grille) || istype(A, /obj/structure/window) || istype(A, /obj/structure/girder))
+		user.Beam(A,icon_state="rped_upgrade",time=20)
+		..()
+	else
+		return
+
+
+
+
+// RAPID LIGHT GENERATOR
+
+
+
+
+
+
+/obj/item/weapon/rld
+	name = "rapid-light-device (RLD)"
+	desc = "A device used to rapidly provide lighting sources to an area."
+	icon = 'icons/obj/tools.dmi'
+	icon_state = "rld-5"
+	opacity = FALSE
+	density = FALSE
+	anchored = FALSE
+	flags = CONDUCT | NOBLUDGEON
+	force = 0
+	throwforce = 10
+	throw_speed = 3
+	throw_range = 5
+	w_class = WEIGHT_CLASS_NORMAL
+	materials = list(MAT_METAL=100000)
+	origin_tech = "engineering=4;materials=2"
+	req_access_txt = "11"
+	armor = list(melee = 0, bullet = 0, laser = 0, energy = 0, bomb = 0, bio = 0, rad = 0, fire = 100, acid = 50)
+	resistance_flags = FIRE_PROOF
+	var/datum/effect_system/spark_spread/spark_system
+	var/matter = 200
+	var/max_matter = 200
+	var/working = 0
+	var/mode = 1
+	actions_types = list(/datum/action/item_action/pick_color)
+
+	var/wallcost = 10
+	var/floorcost = 15
+	var/launchcost = 5
+	var/deconcost = 10
+
+	var/walldelay = 10
+	var/floordelay = 10
+	var/decondelay = 15
+
+	var/sheetmultiplier	= 4
+
+	var/color_choice = null
+
+	var/no_ammo_message = "<span class='warning'>The \'Low Ammo\' light on \
+		the RLD blinks yellow.</span>"
+
+/obj/item/weapon/rld/ui_action_click(mob/user, var/datum/action/A)
+	if(istype(A, /datum/action/item_action/pick_color))
+		choose_color()
+	else
+		..()
+
+/obj/item/weapon/rld/proc/choose_color()
+	set name = "Choose Light Color"
+	set category = "Object"
+	color_choice = input(usr,"Choose Color") as color
+
+
+/obj/item/weapon/rld/New()
+	..()
+
+	desc = "An RLD. It currently holds [matter]/[max_matter] matter-units."
+	src.spark_system = new /datum/effect_system/spark_spread
+	spark_system.set_up(5, 0, src)
+	spark_system.attach(src)
+
+
+/obj/item/weapon/rld/Destroy()
+	qdel(spark_system)
+	spark_system = null
+	. = ..()
+
+/obj/item/weapon/rld/attackby(obj/item/weapon/W, mob/user, params)
```

Comment:
It is but it also uses different materials and changes icons based on how much matter is left so I've found that using a child that has to both add and remove the parent's features tends to leave me with buggy workable solutions.